libraries/WiFi/src/WiFiGeneric.cpp

```diff
@@ -1554,58 +1561,68 @@ bool WiFiGenericClass::setDualAntennaConfig(uint8_t gpio_ant1, uint8_t gpio_ant2
  */
 static void wifi_dns_found_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
 {
+    gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(callback_arg);
     if(ipaddr) {
-        (*reinterpret_cast<IPAddress*>(callback_arg)) = ipaddr->u_addr.ip4.addr;
+        if(parameters->result == 0){
+            memcpy(&(parameters->addr), ipaddr, sizeof(ip_addr_t));
+            parameters->result = 1;
+        }
+    } else {
+        parameters->result = -1;
     }
     xEventGroupSetBits(_arduino_event_group, WIFI_DNS_DONE_BIT);
 }
 
-typedef struct gethostbynameParameters {
-    const char *hostname;
-    ip_addr_t addr;
-    void *callback_arg;
-} gethostbynameParameters_t;
-
 /**
  * Callback to execute dns_gethostbyname in lwIP's TCP/IP context
  * @param param Parameters for dns_gethostbyname call
  */
 static esp_err_t wifi_gethostbyname_tcpip_ctx(void *param)
 {
     gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(param);
-    return dns_gethostbyname(parameters->hostname, &parameters->addr, &wifi_dns_found_callback, parameters->callback_arg);
+    return dns_gethostbyname_addrtype(parameters->hostname, &parameters->addr, &wifi_dns_found_callback, parameters, parameters->addr_type);
 }
 
 /**
- * Resolve the given hostname to an IP address. If passed hostname is an IP address, it will be parsed into IPAddress structure.
- * @param aHostname     Name to be resolved or string containing IP address
+ * Resolve the given hostname to an IP address.
+ * @param aHostname     Name to be resolved
  * @param aResult       IPAddress structure to store the returned IP address
  * @return 1 if aIPAddrString was successfully converted to an IP address,
  *          else error code
  */
-int WiFiGenericClass::hostByName(const char* aHostname, IPAddress& aResult)
+int WiFiGenericClass::hostByName(const char* aHostname, IPAddress& aResult, bool preferV6)
 {
-    if (!aResult.fromString(aHostname))
-    {
-        gethostbynameParameters_t params;
-        params.hostname = aHostname;
-        params.callback_arg = &aResult;
-        aResult = static_cast<uint32_t>(0);
+    err_t err = ERR_OK;
+    gethostbynameParameters_t params;
+
+    aResult = static_cast<uint32_t>(0);
+    params.hostname = aHostname;
+    params.addr_type = preferV6?LWIP_DNS_ADDRTYPE_IPV6_IPV4:LWIP_DNS_ADDRTYPE_IPV4_IPV6;
```

Comment:
Just make this:

params.addr_type = (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT) ? LWIP_DNS_ADDRTYPE_IPV6_IPV4 : LWIP_DNS_ADDRTYPE_IPV4;

If WIFI_WANT_IP6_BIT is turn off, then we only want IPV4. Do not return IPv6 at all, because it might break backwards clients. If a client turns off the IPv6 bit they should only get IPv4 results (like before), i.e. they are IPv4 only. Note the IPV4 only.

If IPv6 is turned on, the return IPv6 if available (thats why they turned it on!), but with fall back to IPv4.

Reply:
This function only returns one address, so if you are doing something where you want both IPv6 global and link-local (although the later don't usually come from DNS) addresses, it isn't very useful.

You really also need to know the context. e.g. if an address has global IPv6, global IPv4, and link-local IPv6, you probably want the global IPv6 (although it is has a relevant link-local maybe use that).

But maybe if you have only link-local IPv6 and a global IPv4, you actually want to return the IPv4.

Reply:
Actually, (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT) is probably not sufficient.

You probably want (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT) && haveGlobalIPv6Address.

Just because you want IPv6 doesn't mean you got it. Maybe you are on an IPv4 only network. You might have auto-assigned yourself an IPv6 link-local address, but if there are no other IPv6 nodes that may be useless, as will a DNS result for google.com that returns you the global IPv6, because you have no way to reach it.

Maybe the network stack takes care of this? (i.e. will it only return addresses you can reach, or will it return all addresses).

RFC 6724 talks about address selection, so may be relevant, where the destination addresses need to be filtered by which make sense based on the available source addresses.

Reply:
yes. it can not be as simple as now, because there are many cases which will cause trouble. I do vote to not bother and change it here much, because after this is merged, I will continue on rewriting the network stack of Arduino and the new things will allow to ask if we have a proper IPv6 address present. What we need to do is define the proper algo to decide what kind of query to issue and how to make sure that results are what they should be.

Reply:
Actually you probably don't even need to check (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT) at all ... all you need to check is (the nonexistent) haveGlobalIPv6Address.

You will only get an address if WANT is on, so re-checking WANT is not necessary.

There are a few different scenarios, if WANT is off, then just do the IPv4 behaviour.  If WANT is on, then sometimes you will be assigned an global/routable address, and then can use IPv6 lookups.  Even if you don't get an address, you can still assign your own link-local, and communicate with other link-local.  DNS probably isn't going to return link-local addresses, but mDNS might.

Happy to go with something initial, and then refine.

I did notice the LWIP has a ip6_select_source_address function based on RFC 6724 (given a particular destination, work out which source address to use from a given interface).

Reply:
Going with the current implementation, you should still change to:

```
 params.addr_type = preferV6?LWIP_DNS_ADDRTYPE_IPV6_IPV4:LWIP_DNS_ADDRTYPE_IPV4;
```

Using `LWIP_DNS_ADDRTYPE_IPV4` instead of `LWIP_DNS_ADDRTYPE_IPV4_IPV6` for when IPv6 is off.

Because if IPv6 is off for backwards compatiblity we only want to return IPv4; otherwise we might break apps (e.g. if no IPv4 we should return nothing instead of IPv6).